File: task/binarysearch/BinarySearch.java
package binarysearch;

public class BinarySearch {

    public static void main(String[] args) {
        System.out.println(search(new int[]{-1, 0, 3, 5, 9, 12}, 9));
        System.out.println(search(new int[]{-1, 0, 3, 5, 9, 12, 16}, 16));
        System.out.println(search(new int[]{2, 5}, 5));
        System.out.println(search(new int[]{-1, 0, 3, 5, 9, 12}, 2));
        System.out.println(search(new int[]{-1, 0, 5}, 5));
    }

    public static int search(int[] nums, int target) {
        if (nums.length == 1) return nums[0] == target ? 0 : -1;
        int mid, left = 0, right = nums.length - 1;

        while (left <= right) {
            mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (target < nums[mid]) {
                right = mid - 1;
            } else left = mid + 1;
        }

        return -1;
    }
}

File: python/week1/BinarySearch.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        right = len(nums) - 1
        left = 0
        while right >= left:
            mid = left + (right - left) // 2
            if nums[mid] > target:
                right = right-1
            elif nums[mid] < target:
                left = left + 1
            else:
                return mid

        return -1



if __name__ == '__main__':
    res1 = Solution().search([-1,0,3,5,9,12], 9)
    res2 = Solution().search([-1,0,3,5,9,12], 2)
    print()